controllers/controller/devworkspacerouting/devworkspacerouting_controller_test.go

```diff
@@ -0,0 +1,504 @@
+// Copyright (c) 2019-2023 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package devworkspacerouting_test
+
+import (
+	"fmt"
+
+	controllerv1alpha1 "github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+	"github.com/devfile/devworkspace-operator/pkg/constants"
+	"github.com/devfile/devworkspace-operator/pkg/infrastructure"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	routeV1 "github.com/openshift/api/route/v1"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+var _ = Describe("DevWorkspaceRouting Controller", func() {
+	Context("Basic DevWorkspaceRouting Tests", func() {
+		It("Gets Preparing status", func() {
```

Comment:
One additional test case that may be useful -- "Does not fail DevWorkspaceRoutings with unrecognized `routingClass`". 

Each DWR reconciler "accepts" certain routingClasses and ignores others -- this is used, for example, in Che to support a custom routing controller. Such a dwr should stay in the "Preparing" phase.

Reply:
Sounds good, will add that test case before this is ready to be merged.

Also, what do you think about a test that involves switching routing classes?

Reply:
Switching routingClass in a dwr isn't supported -- if a DevWorkspace changes its routingClass the routing is [deleted and recreated](https://github.com/devfile/devworkspace-operator/blob/4ea3d5cd8ef0622cf7883b62e7b782941f174b04/pkg/provision/sync/diff.go#L210-L217)

Reply:
> One additional test case that may be useful -- "Does not fail DevWorkspaceRoutings with unrecognized `routingClass`".
> 
> Each DWR reconciler "accepts" certain routingClasses and ignores others -- this is used, for example, in Che to support a custom routing controller. Such a dwr should stay in the "Preparing" phase.

In my testing & investigation, it seems the current functionality is to [stop reconciling](https://github.com/devfile/devworkspace-operator/blob/main/controllers/controller/devworkspacerouting/devworkspacerouting_controller.go#L97) if the routing class is [unrecognized](https://github.com/devfile/devworkspace-operator/blob/main/controllers/controller/devworkspacerouting/solvers/solver.go#L116).

I can either add a commit to address this or make a seperate PR for this :)